style.md

```diff
@@ -811,31 +813,40 @@ seconds that may have occurred between those two instants.
 
 <!-- TODO: section on String methods for enums -->
 
-### Error Types
+### Errors
 
-There are various options for declaring errors:
+#### Error Types
 
-- [`errors.New`] for errors with simple static strings
-- [`fmt.Errorf`] for formatted error strings
-- Custom types that implement an `Error()` method
-- Wrapped errors using [`"pkg/errors".Wrap`]
+There are few options for declaring errors.
+Consider the following before picking the option best suited for your use case.
 
-When returning errors, consider the following to determine the best choice:
+- Does the caller need to match the error so that they can handle it?
+  If yes, we must support the [`errors.Is`] and [`errors.As`] functions
+  by exporting an error variable or type.
+- Is the error message a static string,
+  or is it a dynamic string that requires contextual information?
+  For the former, we can use [`errors.New`], but for the latter we must
+  use [`fmt.Errorf`] or a custom error type.
+- Are we propagating a new error returned by a downstream function?
+  If so, see the [section on error wrapping](#error-wrapping).
 
-- Is this a simple error that needs no extra information? If so, [`errors.New`]
-  should suffice.
-- Do the clients need to detect and handle this error? If so, you should use a
-  custom type, and implement the `Error()` method.
-- Are you propagating an error returned by a downstream function? If so, check
-  the [section on error wrapping](#error-wrapping).
-- Otherwise, [`fmt.Errorf`] is okay.
+[`errors.Is`]: https://golang.org/pkg/errors/#Is
+[`errors.As`]: https://golang.org/pkg/errors/#As
 
-  [`errors.New`]: https://golang.org/pkg/errors/#New
-  [`fmt.Errorf`]: https://golang.org/pkg/fmt/#Errorf
-  [`"pkg/errors".Wrap`]: https://godoc.org/github.com/pkg/errors#Wrap
+| Error matching? | Error Message | Guidance                     |
+|-----------------|---------------|------------------------------|
+| No              | static        | [`errors.New`]               |
+| No              | dynamic       | [`fmt.Errorf`]               |
+| Yes             | static        | variable with [`errors.New`] |
```

Comment:
what do you mean by `variable` here? Did you mean `exported variable`?

In the example below you had
```
var (
  errNotFound = errors.New("not found")
  ErrBrokenLink = errors.New("link is broken")
  ErrCouldNotOpen = errors.New("could not open")
)
```
so all of them are variables?

Reply:
Yeah, they're all variables. Exported or not depends on whether you want to export or not. I'll change the guidance there to say `var` and document the examples further.

Reply:
Yes. I wasn't sure why just one of them calls out `variable`, but others don't.

Reply:
shouldn't the option (1) also be a global variable, just not exported? Avoids an allocation AFAIU ?

Reply:
Depends on re-use. If you're never matching that error, it's a bit annoying and noisy to put every error string into an errors.New.

RE: allocation: Yeah, putting it in a global would save an alloc, but is the inconvenience of pulling every error message out like that worth the savings? I suspect performance sensitive contexts will do that as needed when optimizing their allocations.